### DISCUSSION TITLE
Feature request: Support for note metadata

### DISCUSSION THREAD

**Discussion by Lauloque** [p2or/blender-loom#86] [Ideas]
I love the variables' system for the output file name, I would also love to have the same thing usable in the Metadata's notes input, like this:

![image](https://github.com/p2or/blender-loom/assets/16049822/2b189fe4-e06c-4f91-b84c-7f7eeb494ab5)

Resulting in this:

![image](https://github.com/p2or/blender-loom/assets/16049822/f803f389-6b3f-4ec1-9111-e805e3c796f4)

Even though many of the default variables already exist within the metadata options already, the notes provide a neat customizable one and having Loom's variables in there would make this exponentially more powerful.

**Comment by p2or** [comment]
Hi @L0Lock,

thanks for your request. Great that you came up with this, it's something that has also been on my wish list for a while, and I already did some tests last year. The remaining problem is that I would like to place the data _outside of the frame_, which makes it difficult and takes a lot of time to implement. 

There was a [prototype of sorts from Ubisoft](https://github.com/ubisoft/stampinfo) that is similar to what I have in mind. However the add-on comes with a lot of dependencies and pre-installs that I would like to avoid.

Don't you think it would be more helpful to place the metadata outside the frame? Or wouldn't that matter to you?

Cheers,
Christian

**Comment by Lauloque** [comment]
Glad we share the same interests!

There are many ways to make stamps, some "enlarge" the canvas of the render to put their stuff, some overlay it on the render (Blender), some make a render with a margin to cater for eventual stamps, some have options for all these methods. I think they are all valid, depends the user.

If you want to go as far as building your own stamp system independant from Blender's metadata stamp, it could be entirely done with FFMPEG (at least every company i've worked at used only this far as I know), though that would mean such system isn't usable for still images or image sequences only?

**Comment by p2or** [comment]
Hi @L0Lock,

I think we can have both. For the time being, we can support the built-in system with just a few resources. I created a [new branch](https://github.com/p2or/blender-loom/tree/stamp) to play with:

<img width="445" alt="Screenshot 2024-01-13 at 11 16 52" src="https://github.com/p2or/blender-loom/assets/512368/2402b21e-7e2a-4fd2-b090-c4fb09391d50">

One side effect is that we can support adding new lines which doesn't seem possible out of the box. If you have the time, please download the add-on from there and report back. If you have futher ideas let's just discuss it.

Cheers,
Christian

**Comment by p2or** [comment]
Hi @L0Lock,

I think I've tested every possible case that came to mind, and I'm done so far.  Did you have time to test? Is there anything still bothering you? Or can we merge it in the master?

Cheers,
Christian

**Comment by Lauloque** [comment]
Hi, Sorry for the delay.

It seems to work on my end as well. Just a little possible tweak; It seems to remove some extra spaces.

I tried to manually align my text with spaces, I.E. below I wanted the final text to look as in the text editor:

![image](https://github.com/p2or/blender-loom/assets/16049822/9648febe-c467-4f7a-b61f-cb8908c809a5)

**Comment by p2or** [comment]
Hi @L0Lock,

many thanks for your feedback! 

> Just a little possible tweak; It seems to remove some extra spaces. I tried to manually align my text with spaces, I.E. below I wanted the final text to look as in the text editor...

Great to have you here, [updated the patch](https://github.com/p2or/blender-loom/tree/stamp) to address that:

<img width="699" alt="Screenshot 2024-01-25 at 08 25 53" src="https://github.com/p2or/blender-loom/assets/512368/d1acadee-e568-4a7a-8300-1dfd6552e6e9">

<img width="699" alt="Screenshot 2024-01-25 at 08 44 34" src="https://github.com/p2or/blender-loom/assets/512368/12d9aa5d-efa9-4be4-af47-ba9b76eb7f0c">

Samples:

- `Render Camera:     $CAMERA\nCurrent Frame:     $F4\nFocal Length:      $LENS`
- `Render Camera:     $CAMERA\nCurrent Frame:     $F4\n Focal Length:     $LENS`
- `Render Camera:     $CAMERA\nCurrent Frame:     $F4\nFocale:            $LENS`
- `Render Camera:     $CAMERA\nCurrent Frame:     $F4\n       Focale:     $LENS`



Does that work for you?

Cheers,
Christian

**Reply by Lauloque** [reply]
Hi, sorry again for the delay, got tangled with work ^^' But I tested the branch on my side, and it works like a charm. It even supports conditional variables!

```python
$SAMPLES = bpy.context.scene.eevee.taa_render_samples if bpy.context.scene.render.engine == 'BLENDER_EEVEE_NEXT' else bpy.context.scene.cycles.samples if bpy.context.scene.render.engine == 'CYCLES' else "n/a"
```

![image](https://github.com/user-attachments/assets/f8b8ce90-cad6-4254-8a87-da896fcbf1de)

**Reply by p2or** [reply]
Hi @L0Lock,

Ez... thanks for your tests. Same here, seems to work for all cases. I'll try to merge our branch in the next few days and let you know when I'm done.

Cheers,
Christian

**Reply by p2or** [reply]
Hi @L0Lock,

please apologise the delay - just merged it into the Master branch, finally (dbe479a315bdeb2e4047b322d477185094619d1b). If something still bothers you - I'm open to any criticism, let's just discuss it. Otherwise, have a nice Christmas.

Cheers,
Christian

**Reply by Lauloque** [reply]
Looks good to me :) Have a nice Christmas!
